File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/services/BookingResultRefunding.java
package it.giannibombelli.wsc2026.giftcard.application.services;

import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.giftcard.application.commands.RefundGiftCard;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardRefunding;
import it.giannibombelli.wsc2026.giftcard.domain.policies.RefundGiftCardPolicy;

import static java.util.Objects.requireNonNull;


public class BookingResultRefunding {
    private final RefundGiftCardPolicy policy;
    private final GiftCardRefunding useCase;

    public BookingResultRefunding(RefundGiftCardPolicy refundGiftCardPolicy, GiftCardRefunding giftCardRefunding) {
        this.policy = requireNonNull(refundGiftCardPolicy);
        this.useCase = requireNonNull(giftCardRefunding);
    }

    public void handleBookingResults(BookingResultEvents events) {
        final RefundGiftCard cmd = policy.evaluate(events);

        if (cmd != null) {
            useCase.invoke(cmd);
        }
    }
}

File: typescript/src/giftcard/application/services/bookingResultRefunding.ts
// Servizio applicativo che reagisce ai risultati della prenotazione e rimborsa la gift card.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Booking.

import type { BookingResultEvent } from '@/booking/domain/events/bookingResultEvents.js';
import { RefundGiftCardPolicy } from '../../domain/policies/refundGiftCardPolicy.js';
import { GiftCardRefunding } from '../usecases/giftCardRefunding.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class BookingResultRefunding {
  private readonly _policy: RefundGiftCardPolicy;
  private readonly _useCase: GiftCardRefunding;

  constructor(policy: RefundGiftCardPolicy, useCase: GiftCardRefunding) {
    requireDependency(policy, "policy");
    requireDependency(useCase, "useCase");
    this._policy = policy;
    this._useCase = useCase;
  }

  handleBookingResults(event: BookingResultEvent): void {
    const command = this._policy.evaluate(event);
    if (command !== null) {
      this._useCase.invoke(command);
    }
  }
}
